### PR TITLE
refactor: move init_default_autocmds to lua

### PR DESCRIPTION
**Pull request by glacambre**
The original motivation for this change came from developping
https://github.com/neovim/neovim/pull/22159, which will require adding
more autocommand creation to Neovim's startup sequence.

### CODE REVIEW

File: src/nvim/main.c [hunk 1]
```diff
@@ -415,7 +415,13 @@ int main(int argc, char **argv)
 
   TIME_MSG("init default mappings");
 
-  init_default_autocmds();
+  err = ERROR_INIT;
+  o = NLUA_EXEC_STATIC("return vim._init_default_autocmds()",
```
Comment (bfredl) [inline src/nvim/main.c:419]: I'd suggest merge this with the preceding `NLUA_EXEC_STATIC` call to (1) avoid the repetition of boilerplate (2) entering/exiting the luajit VM is much more expensive than having a lua function just call another lua function.

Comment (glacambre) [reply src/nvim/main.c]: Thanks for the review. I was half-hoping that the diffs I was seeing were coming from a broken setup on my end, but it looks like they're real. I'll turn this PR into a draft, fix it, and then ping you again :)

File: test/functional/api/autocmd_spec.lua [hunk 2]
```diff
@@ -209,7 +209,7 @@ describe('autocmd api', function()
 
       local aus = meths.get_autocmds({ event = 'User', pattern = 'Test' })
       local first = aus[1]
-      eq(first.id, 1)
+      eq(true, first.id > 0)
```
Comment (justinmk) [inline test/functional/api/autocmd_spec.lua:212]: just for future reference: can use ok() to get a nicer failure message.
```suggestion
      ok(first.id > 0, 'first.id > 0', first.id)
```

File: runtime/lua/vim/_editor.lua [hunk 3]
```diff
@@ -894,6 +894,26 @@ function vim._init_default_mappings()
   ]])
 end
 
+function vim._init_default_autocmds()
+  local nvim_terminal_augroup = vim.api.nvim_create_augroup('nvim_terminal', {})
+  vim.api.nvim_create_autocmd({ 'bufreadcmd' }, {
+    pattern = 'term://*',
+    group = nvim_terminal_augroup,
+    nested = true,
+    command = "if !exists('b:term_title')|call termopen(matchstr(expand(\"<amatch>\"), '\\c\\mterm://\\%(.\\{-}//\\%(\\d\\+:\\)\\?\\)\\?\\zs.*'), {'cwd': expand(get(matchlist(expand(\"<amatch>\"), '\\c\\mterm://\\(.\\{-}\\)//'), 1, ''))})",
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:903]: could avoid some escaping by using `[[ ... ]]`

**Comment by justinmk** [comment]
Nice change. Merged to unblock #22159

**Comment by glacambre** [comment]
Thanks a lot for the review!

**Pull request merged**